### DISCUSSION TITLE
Need more details on `ClientTrackingOptions`

### DISCUSSION THREAD

**Discussion by arindamnayak** [redis/rueidis#474] [General]
I see these are possible options for `ClientTrackingOptions`.
- `CLIENT`
- `TRACKING`
- `ON`
- `OPTIN`
- `BCAST`

Can someone explain
-  if the first 4 need to be specified all at one?
- Does bcast only works with prefix?
- Does first 4 ensures that only redis will invalidate the required keys which are queried from a service instance?
- pleas confirm if`BCAST` mode is for receiving  invalidation message from redis server for all the keys? And hence it is better to use with some prefix?
- If above statement is true, then why not use directly use `OPTIN` where redis client handles it automatically by just accepting invalidation message for subscribed keys?

**Comment by rueian** [comment]
Hi @arindamnayak,

> if the first 4 need to be specified all at one?

Yes, you need to specify all client tracking parameters at once when using `ClientTrackingOptions`. However, `CLIENT TRACKING ON OPTIN` is the default used by rueidis, so if this is the case you like to use, then you don't need to specify `ClientTrackingOptions` at all.

> Does bcast only works with prefix?

Yes. See more here: https://redis.io/commands/client-tracking/

> Does first 4 ensures that only redis will invalidate the required keys which are queried from a service instance?

When using the OPTIN mode, redis will only send invalidations of keys that have `CLIENT CACHING YES` preceded. In rueidis, this is done by the `DoCache` and `DoMultiCache`. See https://redis.io/docs/manual/client-side-caching/ for more details.

> pleas confirm ifBCAST mode is for receiving invalidation message from redis server for all the keys? And hence it is better to use with some prefix?

Generally yes.

> If above statement is true, then why not use directly use OPTIN where redis client handles it automatically by just accepting invalidation message for subscribed keys?

Yeap, so rueidis uses OPTIN as default.

**Reply by arindamnayak** [reply]
Thanks @rueian  for the quick clarification.
